worker/remoterelations/remoterelations_test.go

```diff
@@ -0,0 +1,242 @@
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package remoterelations_test
+
+import (
+	"reflect"
+	"time"
+
+	jujutesting "github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	apiservertesting "github.com/juju/juju/apiserver/testing"
+	coretesting "github.com/juju/juju/testing"
+	"github.com/juju/juju/watcher"
+	"github.com/juju/juju/worker"
+	"github.com/juju/juju/worker/remoterelations"
+	"github.com/juju/juju/worker/workertest"
+)
+
+var _ = gc.Suite(&remoteRelationsSuite{})
+
+type remoteRelationsSuite struct {
+	coretesting.BaseSuite
+
+	resources  *common.Resources
+	authorizer *apiservertesting.FakeAuthorizer
+	facade     *mockFacade
+	config     remoterelations.Config
+	stub       *jujutesting.Stub
+}
+
+func (s *remoteRelationsSuite) SetUpTest(c *gc.C) {
+	s.BaseSuite.SetUpTest(c)
+
+	s.stub = new(jujutesting.Stub)
+	s.facade = newMockFacade(s.stub)
+	s.config = remoterelations.Config{
+		Facade: s.facade,
+	}
+}
+
+func (s *remoteRelationsSuite) waitForStubCalls(c *gc.C, expected []jujutesting.StubCall) {
+	var calls []jujutesting.StubCall
+	for a := coretesting.LongAttempt.Start(); a.Next(); {
+		calls = s.stub.Calls()
+		if reflect.DeepEqual(calls, expected) {
+			return
+		}
+	}
+	c.Fatalf("failed to see expected calls. saw: %v", calls)
+}
+
+func (s *remoteRelationsSuite) assertRemoteApplicationWorkers(c *gc.C) worker.Worker {
+	// Checks that the main worker loop responds to remote application events
+	// by starting relevant relation watchers.
+	s.facade.remoteApplications["db2"] = newMockRemoteApplication("db2", "db2url")
+	s.facade.remoteApplications["django"] = newMockRemoteApplication("django", "djangourl")
+	applicationNames := []string{"db2", "django"}
+	s.facade.remoteApplicationsWatcher.changes <- applicationNames
+
+	w, err := remoterelations.New(s.config)
+	c.Assert(err, jc.ErrorIsNil)
+	expected := []jujutesting.StubCall{
+		{"WatchRemoteApplications", nil},
+		{"RemoteApplications", []interface{}{[]string{"db2", "django"}}},
+		{"WatchRemoteApplicationRelations", []interface{}{"db2"}},
+		{"WatchRemoteApplicationRelations", []interface{}{"django"}},
+	}
+	s.waitForStubCalls(c, expected)
+	for _, app := range applicationNames {
+		w, ok := s.facade.remoteApplicationRelationsWatchers[app]
+		c.Check(ok, jc.IsTrue)
+		w.CheckCalls(c, []jujutesting.StubCall{
+			{"Changes", []interface{}{}},
+		})
+	}
+	return w
+}
+
+func (s *remoteRelationsSuite) TestRemoteApplicationWorkers(c *gc.C) {
+	applicationNames := []string{"db2", "django"}
+	w := s.assertRemoteApplicationWorkers(c)
+
+	defer func() {
+		workertest.CleanKill(c, w)
+		// Check that relation watchers are stopped with the worker.
+		for _, app := range applicationNames {
+			w, ok := s.facade.remoteApplicationRelationsWatchers[app]
+			c.Check(ok, jc.IsTrue)
+			for a := coretesting.LongAttempt.Start(); a.Next(); {
```

Comment:
CleanKill waits for the worker to stop, and that should wait for sub-workers to stop. We should fail if it's not killed already?